zerver/lib/rate_limiter.py

```diff
@@ -160,6 +164,106 @@ def rate_limit_entity(cls, entity_key: str, rules: List[Tuple[int, int]],
+        - each key may make atmost `count * (t / window)` request within any t
+          time interval.
+        to
+        - each key may make atmost `count * [(t / window) + 1]` request within
+          any t time interval.
+        Thus, we only need to store reset_times for each key which will be less
+        memory-intensive. This also has the advantage that you can only ever
+        lock yourself out completely for `window / count` seconds instead of
+        `window` seconds.
+        '''
+        now = time.time()
+
+        # Remove all timestamps from `reset_times` that are too old.
+        if cls.last_gc_time <= now - time_window / max_count:
+            cls.last_gc_time = now
+            cls.reset_times = {key: time_dict for (key, time_dict) in cls.reset_times.items()
+                               if time_dict[(time_window, max_count)] > now}
+
+        if entity_key in cls.timestamps_blocked_until:
+            # Check whether the key is manually blocked.
+            if now < cls.timestamps_blocked_until[entity_key]:
+                blocking_ttl = cls.timestamps_blocked_until[entity_key] - now
+                return True, blocking_ttl
+            else:
+                # Remove from timestamps_blocked_until if the blocking
+                # time has been expired.
+                del cls.timestamps_blocked_until[entity_key]
+
+        if entity_key in cls.reset_times:
+            # Check whether the new request would overfill its bucket.
+            new_reset = max(cls.reset_times[entity_key][(time_window, max_count)],
+                            now) + time_window / max_count
+        else:
+            # The bucket was empty, so one request cannot overfill it.
+            new_reset = now + time_window / max_count
+
+        if new_reset > now + time_window:
+            # Compute for how long the bucket will remain filled.
+            time_till_free = new_reset - time_window - now
+            return True, time_till_free
+
+        cls.reset_times[entity_key] = {(time_window, max_count): new_reset}
+        return False, 0.0
+
+    @classmethod
+    def get_api_calls_left(cls, entity_key: str, range_seconds: int,
+                           max_calls: int) -> Tuple[int, float]:
+        now = time.time()
+        reset_time = cls.reset_times[entity_key][(range_seconds, max_calls)]
+        calls_remaining = (now + range_seconds - reset_time) * max_calls // range_seconds
+        return int(calls_remaining), reset_time
+
+    @classmethod
+    def block_access(cls, entity_key: str, seconds: int) -> None:
+        now = time.time()
+        cls.timestamps_blocked_until[entity_key] = now + seconds
+
+    @classmethod
+    def unblock_access(cls, entity_key: str) -> None:
+        del cls.timestamps_blocked_until[entity_key]
+
+    @classmethod
+    def clear_history(cls, entity_key: str) -> None:
+        now = time.time()
+
+        for time_window, max_calls in cls.reset_times[entity_key]:
```

Comment:
I think we need a comment explaining why/what is being done here, because this will be hard to understand for someone who hasn't read through the discussion in #backend